File: samples/cpp/bytetracker_demo.cpp
#include <opencv2/highgui.hpp>
#include <opencv2/dnn.hpp>
#include <opencv2/videoio.hpp>
#include <opencv2/video.hpp>
#include <opencv2/core/utils/logger.hpp>
#include <fstream>

using namespace cv;
using namespace std;
using namespace dnn;

// Yolo 8s detector parameters
const int INPUT_WIDTH = 640;
const int INPUT_HEIGHT = 640;
const float SCORE_THRESHOLD = 0.6f;
const float NMS_THRESHOLD = 0.45f;
const float CONFIDENCE_THRESHOLD = 0.45f;

// Text parameters
const float FONT_SCALE = 0.7f;
const int FONT_FACE = FONT_HERSHEY_SIMPLEX;
const int THICKNESS = 1;

// Colors
const Scalar BLACK = Scalar(0, 0, 0);
const Scalar BLUE = Scalar(255, 178, 50);
const Scalar YELLOW = Scalar(0, 255, 255);
const Scalar RED = Scalar(0, 0, 255);

// Video encoder parameters
const int outputCodec = VideoWriter::fourcc('M', 'J', 'P', 'G');

vector<Mat> inference(Mat&, Net&);
Mat postProcessImage(Mat&, vector<Mat>&, const vector<string>&, vector<Detection>&);
void drawLabel(Mat&, string, int, int);
void writeDetectionsToFile(const vector<Detection>, const string&, const int frameNumber);
void writeTracksToFile(const Mat&, const string&, const int frameNumber);
void writeTracksToFile(const vector<Track> objects, const string &outputPath, const int frameNumber);
Scalar getColor(int);
Mat detectionToMat(vector<Detection>);

const char *keys =
        "{ help     h  |   | Print help message }"
        "{ input    i  |   | Full path to input video folder, the specific camera index. (empty for camera 0) }"
        "{ video    v  | output.avi | Path to video file for recording recording }"
        "{ net         | yolov8s.onnx | Yolo8s detector model }"
        "{ classes     |   | Optional path to a text file with names of classes to label detected objects. }"
        "{ dets_path   |   | Optional path to store all detected boxes }"
        "{ tracks_path |   | Optional path to store all tracks }"
        "{ backend     | 0 | Choose one of computation backends: "
                            "0: automatically (by default), "
                            "1: Halide language (http://halide-lang.org/), "
                            "2: Intel's Deep Learning Inference Engine (https://software.intel.com/openvino-toolkit), "
                            "3: OpenCV implementation, "
                            "4: VKCOM, "
                            "5: CUDA },"
        "{ target      | 0 | Choose one of target computation devices: "
                            "0: CPU target (by default), "
                            "1: OpenCL, "
                            "2: OpenCL fp16 (half-float precision), "
                            "3: VPU, "
                            "4: Vulkan, "
                            "6: CUDA, "
                            "7: CUDA fp16 (half-float preprocess) }";

int main(int argc, char** argv)
{
    CommandLineParser parser(argc, argv, keys);

    if (parser.has("help"))
    {
        parser.printMessage();
        return 0;
    }

    // Load class list.
    vector<string> classList;
    if (parser.has("classes"))
    {
        ifstream ifs(parser.get<std::string>("classes"));
        string line;
        while (getline(ifs, line))
        {
            classList.push_back(line);
        }
    }

    std::string inputVideo = parser.get<std::string>("input");
    VideoCapture capture(inputVideo);
    if (!capture.isOpened())
    {
        cout << "failed to open video: " << inputVideo << endl;
        return -1;
    }

    // Load model.
    std::string net = parser.get<std::string>("net");
    int backend = parser.get<int>("backend");
    int target = parser.get<int>("target");
    Net detector = readNet(net);
    detector.setPreferableBackend(backend);
    detector.setPreferableTarget(target);
    vector<Mat> detections; // Process the image.
    vector<Detection> objects;

    double fps = capture.get(CAP_PROP_FPS);
    int width = capture.get(CAP_PROP_FRAME_WIDTH);
    int height = capture.get(CAP_PROP_FRAME_HEIGHT);

    cv::ByteTracker::Params params;
    params.frameRate = fps;
    params.frameBuffer = 30;
    Ptr<ByteTracker> tracker = ByteTracker::create(params);

    std::string outputVideo = parser.get<std::string>("video");
    VideoWriter writer(outputVideo, outputCodec, fps, cv::Size(width, height));

    std::string detectorOutputPath;
    if(parser.has("dets_path"))
        detectorOutputPath = parser.get<std::string>("dets_path");

    std::string trackerOutputPath;
    if(parser.has("tracks_path"))
        trackerOutputPath = parser.get<std::string>("tracks_path");

    Mat frame;
    int frameNumber = 0;
    while (capture.read(frame))
    {
        if (frame.empty())
        {
            cout << "Failed to read the frame." << endl;
            break;
        }

        detections = inference(frame, detector);
        Mat img = postProcessImage(frame, detections, classList, objects);

        std::cout << "detections for " << frameNumber << ":" << std::endl;
        for(size_t i = 0; i < objects.size(); i++)
            std::cout << objects[i].rect << ", " << objects[i].classLabel << ", " << objects[i].classScore << std::endl;

        vector<Track> trackedObjects;
        tracker->update(objects, trackedObjects); //Update method with vector of detection as input and vector of tracks as output
        for (auto& track : trackedObjects)
        {
            Scalar color = getColor(track.trackingId);
            rectangle(img, track.rect, color, 2);
            int xPoint = static_cast<int>(track.rect.x);
            int yPoint = static_cast<int>(track.rect.y);
            putText(img, to_string(track.trackingId), Point(xPoint, yPoint - 5), FONT_FACE, FONT_SCALE, RED);
        }
        if(!trackerOutputPath.empty())
            writeTracksToFile(trackedObjects, trackerOutputPath, frameNumber);

        if(!detectorOutputPath.empty())
            writeDetectionsToFile(objects, detectorOutputPath, frameNumber);

        vector<double> layersTimes;
        double freq = getTickFrequency() / 1000;
        double t = detector.getPerfProfile(layersTimes) / freq;
        string label = format("Inference time : %.2f ms", t);
        putText(img, label, Point(20, 40), FONT_FACE, FONT_SCALE, RED);

        imshow("Output", img);
        writer.write(img);
        //waitKey(0);

        if (waitKey(1) == 27)
            break;

        frameNumber++;
    }
    writer.release();
    capture.release();

    return 0;
}

vector<Mat> inference(Mat &inputImage, Net &net)
{
    // Convert to blob.
    Mat blob;
    dnn::blobFromImage(inputImage, blob, 1. / 255., Size(INPUT_WIDTH, INPUT_HEIGHT), Scalar(), true, false);

    net.setInput(blob);

    // Forward propagate.
    vector<Mat> outputs;
    net.forward(outputs, net.getUnconnectedOutLayersNames());

    return outputs;
}

Mat postProcessImage(Mat &inputImage, vector<Mat> &output, const vector<string> &className,
    vector<Detection> &object)
{
    // Initialize vectors to hold respective outputs while unwrapping detections.
    vector<int> classIds;
    vector<float> confidences;
    vector<Rect> boxes;
    //cout << "Original Height and Width :" << inputImage.rows << "x" << inputImage.cols << endl;
    // Resizing factor.
    float xFactor = static_cast<float>(1.0f * inputImage.cols / INPUT_WIDTH);
    float yFactor = static_cast<float>(1.0f * inputImage.rows / INPUT_HEIGHT);
    bool yolov8 = false;

    int rows = output[0].size[1];
    int dimensions = output[0].size[2];

    if (dimensions > rows) // Check if shape[2] is more than shape[1] (yolov8)
    {
        yolov8 = true;
        rows = output[0].size[2];
        dimensions = output[0].size[1];
        output[0] = output[0].reshape(1, dimensions);
        cv::transpose(output[0], output[0]);
    }

    // Iterate through detections.
    float *data = (float*)output[0].data;
    Point classId;
    double maxClassScore;
    for (int i = 0; i < rows; ++i)
    {
        float confidence;
        float *classes_scores = 0;
        if (yolov8)
        {
            classes_scores = data + 4;
        }
        else
        {
            confidence = data[4];
            // Discard bad detections and continue.
            if (confidence >= CONFIDENCE_THRESHOLD)
            {
                classes_scores = data + 5;
            }
        }
        Mat scores(1, 80, CV_32FC1, classes_scores); //80 classes
        // Perform minMaxLoc and acquire the index of best class  score.
        minMaxLoc(scores, 0, &maxClassScore, 0, &classId);
        // Continue if the class score is above the threshold.
        if (maxClassScore > SCORE_THRESHOLD)
        {
            // Store class ID and confidence in the pre-defined respective vectors.
            if (yolov8)
            {
                confidences.push_back(static_cast<float>(maxClassScore));
            }
            else
            {
                confidences.push_back(confidence);
            }

            classIds.push_back(classId.x);
            // Center.
            float cx = data[0];
            float cy = data[1];
            // Box dimension.
            float w = data[2];
            float h = data[3];
            // Bounding box coordinates.
            int left = int((cx - 0.5 * w) * xFactor);
            int top = int((cy - 0.5 * h) * yFactor);
            int width = int(w * xFactor);
            int height = int(h * yFactor);
            // Store good detections in the boxes vector.
            boxes.push_back(Rect(left, top, width, height));
        }
        // Jump to the next row.
        data += dimensions;
    }
    vector<int> nmsResult;
    dnn::NMSBoxes(boxes, confidences, SCORE_THRESHOLD, NMS_THRESHOLD, nmsResult);
    for (size_t i = 0; i < nmsResult.size(); i++)
    {
        int idx = nmsResult[i];
        Rect box = boxes[idx];
        int top = box.y;
        int left = box.x;
        int width = box.width;
        int height = box.height;
        // Draw bounding box.
        rectangle(inputImage, Point(left, top), Point(left + width, top + height), BLUE, 1 * THICKNESS);
        // Get the label for the class name and its confidence.
        string label = format("%.2f", confidences[idx]);
        label = className.empty() ? cv::format("class %d", classIds[idx]): className[classIds[idx]] + ":" + label;
        // Draw class labels.
        drawLabel(inputImage, label, left, top);
        Detection det(box, classIds[idx], confidences[idx]);
        object.push_back(det);
    }

    return inputImage;
}

void drawLabel(Mat &inputImage, string label, int left, int top)
{
    // Display the label at the top of the bounding box.
    int baseLine;
    Size labelSize = getTextSize(label, FONT_FACE, FONT_SCALE, THICKNESS, &baseLine);
    top = max(top, labelSize.height);
    // Top left corner.
    Point tlc = Point(left, top);
    // Bottom right corner.
    Point brc = Point(left + labelSize.width, top + labelSize.height + baseLine);
    // Draw white rectangle.
    rectangle(inputImage, tlc, brc, BLACK, FILLED);
    // Put the label on the black rectangle.
    putText(inputImage, label, Point(left, top + labelSize.height), FONT_FACE, FONT_SCALE, YELLOW, THICKNESS);
}

void writeDetectionsToFile(const vector<Detection> objects, const string &outputPath,
    const int frameNumber)
{
    // Open the output file for writing
    ofstream outputFile(outputPath, ios_base::app);
    if (!outputFile.is_open())
    {
        cout << "Failed to open the output file: " << outputPath << endl;
        return;
    }
    // Check if the output file is empty
    ifstream file(outputPath);
    bool isEmpty = file.peek() == ifstream::traits_type::eof();
    // Write the header row of the file is empty
    if (isEmpty)
    {
        outputFile << "frame, trackId, x, y, width, height, score, classId" << endl;
    }

    // Iterate over the detections and write the data to the output file
    for (const auto& object : objects)
    {
        // Extract the detection data (frame, trackId, x, y, width, height, score, classId)
        float y = object.rect.y;
        float x = object.rect.x;
        float width = object.rect.width;
        float height = object.rect.height;
        float score = object.classScore;
        int classId = object.classLabel;

        // Write the data to the output file
        outputFile << frameNumber << ", " << -1 << ", " << x << ", " << y << ", " << width << ", " << height << ", " << score << ", "<<classId << endl;
    }

    // Close the output file
    outputFile.close();

    //cout << "\n Detection data saved to: " << outputPath << endl;
}

void writeTracksToFile(const Mat& objects, const string &outputPath,
    const int frameNumber)
{
    // Open the output file for writing
    ofstream outputFile(outputPath, ios_base::app);
    if (!outputFile.is_open())
    {
        cout << "Failed to open the output file: " << outputPath << endl;
        return;
    }
    // Check if the output file is empty
    ifstream file(outputPath);
    bool isEmpty = file.peek() == ifstream::traits_type::eof();
    // Write the header row of the file is empty
    if (isEmpty)
    {
        outputFile << "frame, trackId, x, y, width, height, score, classId" << endl;
    }

    // Iterate over the detections and write the data to the output file
    for (int i = 0; i < objects.rows; ++i)
    {
        // Extract the detection data (frame, trackId, x1, y1, width, height, score, classId)
        float x = objects.at<float>(i, 0);
        float y = objects.at<float>(i, 1);
        float width = objects.at<float>(i, 2);
        float height = objects.at<float>(i, 3);
        int classId = static_cast<int>(objects.at<float>(i, 4));
        float score = objects.at<float>(i, 5);
        int trackId = static_cast<int>(objects.at<float>(i, 6));

        // Write the data to the output file
        outputFile << frameNumber << ", " << trackId << ", " << x << ", " << y << ", " << width << ", " << height << ", " << score << ", " << classId << endl;
    }

    // Close the output file
    outputFile.close();
    //cout << "\n Detection data saved to: " << outputPath << endl;
}

void writeTracksToFile(const vector<Track> objects, const string &outputPath,
    const int frameNumber)
{
    // Open the output file for writing
    ofstream outputFile(outputPath, ios_base::app);
    if (!outputFile.is_open())
    {
        cout << "Failed to open the output file: " << outputPath << endl;
        return;
    }
    // Check if the output file is empty
    ifstream file(outputPath);
    bool isEmpty = file.peek() == ifstream::traits_type::eof();
    // Write the header row of the file is empty
    if (isEmpty)
    {
        outputFile << "frame, trackId, x, y, width, height, score, classId" << endl;
    }


    // Iterate over the detections and write the data to the output file
    for (auto track : objects)
    {
        // Extract the detection data (frame, trackId, x1, y1, width, height, score, classId)
        float x = track.rect.x;
        float y = track.rect.y;
        float width = track.rect.width;
        float height = track.rect.height;
        int classId = track.classLabel;
        float score = track.classScore;
        int trackId = track.trackingId;


        // Write the data to the output file
        outputFile << frameNumber << ", " << trackId << ", " << x << ", " << y << ", " << width << ", " << height << ", " << score << ", " << classId << endl;
    }
    // Close the output file
    outputFile.close();
    //cout << "\n Detection data saved to: " << outputPath << endl;
}

Scalar getColor(const int idx)
{
    int value = idx + 3;
    return Scalar(37 * value % 255, 17 * value % 255, 29 * value % 255);
}

Mat detectionToMat(vector<Detection> objs)
{
    Mat output(static_cast<int>(objs.size()), 6, CV_32F);
    for (int i = 0; i < static_cast<int>(objs.size()); ++i)
    {
        const Detection& detection = objs[i];
        cv::Mat row = output.row(i);

        row.at<float>(0) = detection.rect.x;
        row.at<float>(1) = detection.rect.y;
        row.at<float>(2) = detection.rect.width;
        row.at<float>(3) = detection.rect.height;
        row.at<float>(4) = static_cast<float>(detection.classLabel);
        row.at<float>(5) = detection.classScore;
    }

    return output;
}

File: samples/python/bytetracker_demo.py
#import cv2
import numpy as np
import os
import sys

import site; print(site.getsitepackages())

# Constants
INPUT_WIDTH = 640
INPUT_HEIGHT = 640
SCORE_THRESHOLD = 0.5
NMS_THRESHOLD = 0.45
CONFIDENCE_THRESHOLD = 0.45

FONT_SCALE = 0.7
THICKNESS = 1

# Colors
BLACK = (0, 0, 0)
BLUE = (255, 178, 50)
YELLOW = (0, 255, 255)
RED = (0, 0, 255)

home = os.getenv("HOME")
DETECTIONS_OUTPUT_PATH = os.path.join(home, "files", "det_python.txt")
TRACKINGS_OUTPUT_PATH = os.path.join(home, "files", "tracked_python.txt")
VIDEO_OUTPUT_PATH = os.path.join(home, "files", "output_python.mp4")
COCO_NAMES = os.path.join(home, "files", "coco.names")
#NET_PATH = os.path.join(home, "files", "yolov5s.onnx")
NET_PATH = os.path.join(home, "files", "yolov8s.onnx")
#NET_PATH = os.path.join(home, "files", "yolov8x.onnx")
#NET_PATH = os.path.join(home, "files", "bytetrack_x_mot20.onnx")

PYTHON_PATH = os.path.join(home,"opencv","python","bindings")

sys.path.append(PYTHON_PATH)
import cv2

#print(dir(cv2))
#print(sys.path)

FONT_FACE = cv2.FONT_HERSHEY_SIMPLEX

output_codec = cv2.VideoWriter_fourcc(*'XVID')
output_fps = 10
output_size = (768, 576)

class Detection:
    def __init__(self, rect, class_id, confidence):
        self.rect = rect
        self.class_label = class_id
        self.class_score = confidence



def get_color(idx):
    value = idx + 3
    return (int(37 * value % 255), int(17 * value % 255), int(29 * value % 255))

def drawLabel(input_image, label, left, top):
    font_scale = 0.7
    font_face = cv2.FONT_HERSHEY_SIMPLEX
    thickness = 1
    base_line = 0

    label_size, _ = cv2.getTextSize(label, font_face, font_scale, thickness)
    top = max(top, label_size[1])

    tlc = (left, top)
    brc = (left + label_size[0], top + label_size[1] + base_line)

    cv2.rectangle(input_image, tlc, brc, BLACK, cv2.FILLED)
    cv2.putText(input_image, label, (left, top + label_size[1]), font_face, font_scale, YELLOW, thickness)

def pre_process_image(input_image, net):
    blob = cv2.dnn.blobFromImage(input_image, 1.0 / 255.0, (INPUT_WIDTH, INPUT_HEIGHT), swapRB=True, crop=False)
    net.setInput(blob)
    outputs = net.forward(net.getUnconnectedOutLayersNames())
    return outputs

def post_process_image(input_image, output, class_names, objects):
    class_ids = []
    confidences = []
    boxes = []

    x_factor = input_image.shape[1] / INPUT_WIDTH
    y_factor = input_image.shape[0] / INPUT_HEIGHT
    yolov8 = False

    rows = output[0].shape[1]
    dimensions = output[0].shape[2]

    if dimensions > rows:
        yolov8 = True
        rows = output[0].shape[2]
        dimensions = output[0].shape[1]
        #84x8400 to rows of 1x84 (8400 rows)
        new_output = output[0]
        new_output = new_output.T
        #print(output)
        output = (new_output,)
        #print(output[0].shape)

    #Iterate through 25200 detections.
    data = output[0]
    #print(len(data))
    max_class_score = 0.0

    for i in range(rows):
        #print(len(data[i]))
        if yolov8:
            #print(data[0])
            classes_scores = data[i][4:]
            #print(len(classes_scores))
            class_id = np.argmax(classes_scores)
            max_class_score = classes_scores[class_id][0]
            if max_class_score > SCORE_THRESHOLD:

                confidences.append(max_class_score)
                class_ids.append(class_id)

                x = data[i][0]
                y = data[i][1]
                w = data[i][2]
                h = data[i][3]

                left = int((x - 0.5 * w) * x_factor)
                top = int((y - 0.5 * h) * y_factor)
                width = int(w * x_factor)
                height = int(h * y_factor)

                boxes.append((left, top, width, height))
        else:
            confidence = data[i][4]
            if confidence >= CONFIDENCE_THRESHOLD:
                classes_scores = data[i][5:]
                class_id = np.argmax(classes_scores)
                max_class_score = classes_scores[class_id]

                if max_class_score > SCORE_THRESHOLD:
                    confidences.append(confidence)
                    class_ids.append(class_id)

                    cx = data[i][0]
                    cy = data[i][1]
                    w = data[i][2]
                    h = data[i][3]

                    left = int((cx - 0.5 * w) * x_factor)
                    top = int((cy - 0.5 * h) * y_factor)
                    width = int(w * x_factor)
                    height = int(h * y_factor)

                    boxes.append((left, top, width, height))

        #data = data[dimensions:]

    #print(confidences)
    nms_indices = cv2.dnn.NMSBoxes(boxes, confidences, SCORE_THRESHOLD, NMS_THRESHOLD)
    #print(nms_indices)

    #print("max idx",max(nms_indices))
    for i in nms_indices:
        box = boxes[i] #check this part
        #print(box)
        left, top, width, height = box
        class_id = class_ids[i]
        #class_name = class_names[class_id]
        confidence = confidences[i]

        #color = get_color(class_id)
        #cv2.rectangle(input_image, (left, top), (left + width, top + height), color, 2)
        #label = f"{class_name}: {confidence:.2f}"
        #drawLabel(input_image, label, left, top)

        det = Detection((left, top, width, height), class_id, confidence)
        objects.append(det)

    return input_image

def detection_to_mat(objs):
    output = np.zeros((len(objs), 6), dtype=np.float32)

    for i, det in enumerate(objs):
        output[i, 0] = det.rect[0]
        output[i, 1] = det.rect[1]
        output[i, 2] = det.rect[2]
        output[i, 3] = det.rect[3]
        output[i, 4] = det.class_label
        output[i, 5] = det.class_score

    return output

def write_tracks_to_file(tracked_objects, output_path, frame_number):
    with open(output_path, "a") as output_file:
        if output_file.tell() == 0:
            output_file.write("frame, trackId, x, y, width, height, score, classId\n")

        for i in range(tracked_objects.shape[0]):
            x = int(tracked_objects[i, 0])
            y = int(tracked_objects[i, 1])
            width = int(tracked_objects[i, 2])
            height = int(tracked_objects[i, 3])
            class_id = int(tracked_objects[i, 4])
            score = round(tracked_objects[i, 5],3)
            track_id = int(tracked_objects[i, 6])

            output_file.write(f"{frame_number}, {track_id}, {x}, {y}, {width}, {height}, {score}, {class_id}\n")

def write_detections_to_file(objects, output_path, frame_number):
    with open(output_path, "a") as output_file:
        if output_file.tell() == 0:
            output_file.write("frame, trackId, x, y, width, height, score, classId\n")

        for det in objects:
            x = int(det.rect[0])
            y = int(det.rect[1])
            width = int(det.rect[2])
            height = int(det.rect[3])
            class_id = det.class_label
            score = round(det.class_score,3)

            output_file.write(f"{frame_number}, {-1}, {x}, {y}, {width}, {height}, {score}, {class_id}\n")





def main():
    # Load class list
    class_list = []
    with open(COCO_NAMES, 'r') as f:
        class_list = [line.strip() for line in f.readlines()]

    # Load video capture
    capture = cv2.VideoCapture("../data/vtest.avi")
    if not capture.isOpened():
        print("Failed to open the video.")
        return

    # Load tracker
    params = cv2.ByteTracker_Params()
    params.frameRate = int(capture.get(cv2.CAP_PROP_FPS))
    params.frameBuffer = 30
    tracker = cv2.ByteTracker_create(params)

    # Create video writer
    fourcc = cv2.VideoWriter_fourcc(*'XVID')
    writer = cv2.VideoWriter(VIDEO_OUTPUT_PATH, fourcc, output_fps, output_size)

    frame_number = 0
    while True:
        ret, frame = capture.read()
        if not ret:
            break

        # Load model
        net = cv2.dnn.readNet(NET_PATH)

        # Pre-process image
        detections = pre_process_image(frame, net)

        # Post-process image
        objects = []
        #frame_clone = frame.copy()
        img = post_process_image(frame, detections, class_list, objects)
        objects_mat = detection_to_mat(objects)
        ok, tracked_objects = tracker.update(objects_mat) #I need to hardcode the output

        if ok:
            for i in range(tracked_objects.shape[0]):
                #print(tracked_objects)
                id_ = int(tracked_objects[i, 6])
                color = get_color(id_)
                tlwh_ = (
                    tracked_objects[i, 0],
                    tracked_objects[i, 1],
                    tracked_objects[i, 2],
                    tracked_objects[i, 3]
                )

                cv2.rectangle(img, (int(tlwh_[0]), int(tlwh_[1])),
                            (int(tlwh_[0] + tlwh_[2]), int(tlwh_[1] + tlwh_[3])),
                            color, 2)
                cv2.putText(img, str(id_), (int(tlwh_[0]), int(tlwh_[1]) - 5),
                            FONT_FACE, FONT_SCALE, RED, THICKNESS)

            write_tracks_to_file(tracked_objects, TRACKINGS_OUTPUT_PATH, frame_number)
            write_detections_to_file(objects, DETECTIONS_OUTPUT_PATH, frame_number)

            layers_times = net.getPerfProfile()[0] / cv2.getTickFrequency()
            label = f"Inference time: {layers_times * 1000:.2f} ms"
            cv2.putText(img, label, (20, 40), FONT_FACE, FONT_SCALE, RED, THICKNESS)

            cv2.imshow("Output", img)
            writer.write(img)

            if cv2.waitKey(1) == 27:
                break

            frame_number += 1

    writer.release()
    capture.release()
    print("Output video generated")

if __name__ == "__main__":
    main()
